tgui/packages/tgui/interfaces/PreferencesMenu/KeybindingsPage.tsx

```diff
@@ -0,0 +1,473 @@
+    );
+
+    moveToBottom(keybindingEntries, "EMOTE");
+    moveToBottom(keybindingEntries, "ADMIN");
+
+    return (
+      <>
+        <KeyListener
+          onKeyDown={this.handleKeyDown}
+          onKeyUp={this.handleKeyUp}
+        />
+
+        <TabbedMenu
+          categoryEntries={keybindingEntries.map(
+            ([category, keybindings]) => {
+              return [category, (
+                <Stack key={category} vertical fill>
+                  {sortKeybindings(Object.entries(keybindings)).map(
+                    ([keybindingId, keybinding]) => {
+                      const keys
+                        = this.state.selectedKeybindings![keybindingId]
+                          || [];
+
+                      const name = (
+                        <Stack.Item basis="25%">
+                          <KeybindingName keybinding={keybinding} />
+                        </Stack.Item>
+                      );
+
+                      return (
+                        <Stack.Item key={keybindingId}>
+                          <Stack fill>
+                            {name}
+
+                            <Stack.Item grow basis="10%">
+                              <KeybindingButton
+                                currentHotkey={keys[0]}
+                                typingHotkey={this.getTypingHotkey(
+                                  keybindingId,
+                                  0,
+                                )}
+                                onClick={this.getKeybindingOnClick(
+                                  keybindingId,
+                                  0,
+                                )}
+                              />
+                            </Stack.Item>
+
+                            <Stack.Item grow basis="10%">
+                              <KeybindingButton
+                                currentHotkey={keys[1]}
+                                typingHotkey={this.getTypingHotkey(
+                                  keybindingId,
+                                  1,
+                                )}
+                                onClick={this.getKeybindingOnClick(
+                                  keybindingId,
+                                  1,
+                                )}
+                              />
+                            </Stack.Item>
+
+                            <Stack.Item grow basis="10%">
+                              <KeybindingButton
+                                currentHotkey={keys[2]}
+                                typingHotkey={this.getTypingHotkey(
+                                  keybindingId,
+                                  2,
+                                )}
+                                onClick={this.getKeybindingOnClick(
+                                  keybindingId,
+                                  2,
+                                )}
+                              />
+                            </Stack.Item>
```

Comment:
you could probably shorten this code a lot by moving it to a functional component or using a for loop. A lot of it is the same.